Implement Redis.Scan

brian-carroll opened the pull request.

jwoudenberg-nri commented on nri-prelude/src/Test/Internal.hs:
```diff
@@ -150,14 +150,21 @@ todo name =
+test = Stack.withFrozenCallStack mkTest defaultTimeout
+
+-- | Same as test, but with increased timeout.
+slow :: Stack.HasCallStack => Int -> Text -> (() -> Expectation) -> Test
```
💭 Maybe use a `Float` to represent milliseconds, because other time-taking Elm-like API's like `sleep` and `every` use `Float` in milliseconds as a quasi-standard for durations.

jwoudenberg-nri commented on nri-prelude/tests/golden-results/test-report-stdout-tests-failed:
```diff
@@ -19,7 +19,7 @@ Called from:
 ↓ sub suite
 ✗ test 7
-Test timed out
+Test timed out after 10000
```
😻 This is useful extra information! We should add a unit!

jwoudenberg-nri commented on nri-prelude/src/Test/Internal.hs:
```diff
@@ -50,7 +50,7 @@ data TestResult
   = FailedAssertion Text (Maybe Stack.SrcLoc)
   | ThrewException Exception.SomeException
-  | TookTooLong
+  | TookTooLong Int
```
😻

brian-carroll committed:
Implement SCAN command

brian-carroll committed:
Factor out Cursor encoding and decoding

brian-carroll committed:
Use LibraryError

brian-carroll committed:
Allow Redis.Mock to store scan state

The real Redis does seem to store some state during the full scan,
because we can delete keys between individual scan calls.

brian-carroll committed:
Ensure Mock.Redis.scan will work when used to delete keys

brian-carroll committed:
correct comments in NonEmptyDict

brian-carroll committed:
Add a test for scan

brian-carroll committed:
Remove namespace prefix from keys returned by SCAN

brian-carroll committed:
Fix scan termination bug in Redis.Mock

brian-carroll committed:
Get rid of ScanId newtype (it was always immediately unwrapped)

brian-carroll committed:
Move Redis.foldWithScan from tests into library

brian-carroll commented on nri-redis/nri-redis.cabal:
```diff
@@ -71,6 +71,7 @@ library
     , nri-env-parser >=0.1.0.0 && <0.2
     , nri-observability >=0.1.0 && <0.2
     , nri-prelude >=0.1.0.0 && <0.7
+    , pcre-light >=0.4.1.0 && <0.4.2
```
Used in Redis.Mock to implement scan patterns

brian-carroll commented on nri-redis/src/Redis/Mock.hs:
```diff
@@ -47,21 +49,18 @@ handlerIO = do
 data Model = Model
-  {hash :: HM.HashMap Text RedisType}
+  { hash :: HM.HashMap Text RedisType,
+    scans :: Scans
```
We need some internal state to implement a `scan` that works correctly in the case where we are actually _deleting_ keys while scanning them! Using the cursor as the only form of state doesn't work (or at least I couldn't see a way)

brian-carroll commented on nri-redis/src/Redis/Mock.hs:
```diff
@@ -106,299 +105,413 @@ expectInt val =
+            ( \key (hm', count) ->
+                if HM.member key hm'
+                  then (HM.delete key hm', count + 1)
+                  else (hm', count)
```
GitHub is rendering this diff in the most horrible way. Not sure what I can do about it.
Basically all branches now need to update the whole `model` instead of just `hm`.

brian-carroll commented on nri-redis/src/Redis/Internal.hs:
```diff
@@ -354,3 +417,21 @@ traceQuery commands host port task =
+        if nextCursor == Database.Redis.cursor0
+          then Task.succeed nextAccumulator
+          else go nextAccumulator nextCursor
+   in go initAccumulator Database.Redis.cursor0
```
I'm not sure how well `foldWithScan` fits with the overall API.

I experimented with creating a monad instance instead, but got stuck implementing `cmds` and `mapKeys`. Not sure how to do those without actually running the queries.

In my first attempt I tried creating an `AndThen` constructor. Got stuck, tried with a `Join` constructor instead, got stuck again. Along the way, I realised that `Monad` creates a lot of ways of composing things, and all of that needs to be tested. It simplifies matters quite a bit to just directly support the exact thing I need.

Also, running it in Task allows you to do anything you like with the keys, not just other Redis queries.

stoeffel commented on nri-redis/src/Redis.hs:
```diff
@@ -116,6 +120,12 @@ data Api key a = Api
     --
     -- https://redis.io/commands/ping
     ping :: Internal.Query (),
+    -- | You should probably use `foldWithScan` instead of this function.
```
Seeing how complex this is, I wonder if adding `keys` wouldn't be simpler and also solve your use-case.

brian-carroll replied:
Hmm well I felt the only really complex part was `Redis.Mock`.

And Redis seem to recommend against `keys` in production

https://redis.io/commands/keys/

> Warning: consider KEYS as a command that should only be used in production environments with extreme care. It may ruin performance when it is executed against large databases. This command is intended for debugging and special operations, such as changing your keyspace layout. Don't use KEYS in your regular application code. If you're looking for a way to find keys in a subset of your keyspace, consider using [SCAN](https://redis.io/commands/scan) or [sets](https://redis.io/topics/data-types#sets).


https://redis.io/commands/scan/

> Since these commands allow for incremental iteration, returning only a small number of elements per call, they can be used in production without the downside of commands like [KEYS](https://redis.io/commands/keys) or [SMEMBERS](https://redis.io/commands/smembers) that may block the server for a long time (even several seconds) when called against big collections of keys or elements.

stoeffel replied:
yeah, I was just considering keys because it's for a mostly dev feature right?

stoeffel replied:
Should we only expose `foldWithScan`?

brian-carroll replied:
Yeah I was debating that too and wondered what you'd think.
Yes I think it's probably better that way.
I'll delete `scan`.

brian-carroll replied:
> yeah, I was just considering keys because it's for a mostly dev feature right?

Right, I see what you mean.
We do have a "clear content cache" button in prod though, so it should probably be safe to press it.
I know Curriculum sometimes complains about the cache delay. I wonder if they or QA sometimes clear it?
And once we expose `keys` from the library, people can use it for whatever.

stoeffel reviewed (approved):
I think this looks good!
The only thing I wonder is if we should only expose `foldWithScan` since it's the recommended usage.

brian-carroll committed:
Remove low-level `scan` command, since `foldWithScan` is recommended

brian-carroll commented:
OK that's done.
It also allowed me to un-expose the `Cursor` type and `cursor0` value. Those are now internal details, which is great.
This definitely feels better.
Thanks for the review.

stoeffel reviewed (approved):
LGTM!
thanks!
